File: alert_handler.go
package main

type alertHandler struct {
	repository deviceRepositoryPort
}

func (h alertHandler) onNetworkInterfaceAlert(sourceDeviceName string, targetDeviceName string) {
	sourceDevice := h.getDevice(sourceDeviceName)
	targetDevice := h.getDevice(targetDeviceName)
	sourceDevice.alertsCount += 1
	targetDevice.alertsCount += 1
	h.updateDevice(sourceDeviceName, sourceDevice, targetDevice)
	h.updateDevice(targetDeviceName, targetDevice, sourceDevice)
}

func (h alertHandler) getDevice(deviceName string) Device {
	device := h.repository.findDevice(deviceName)
	// Don't know how to deal with 'Optional' results...
	if device.name == "" {
		device.name = deviceName
	}
	return device
}

func (h alertHandler) updateDevice(deviceName string, device Device, neighbour Device) {
	h.repository.updateAlertsCount(deviceName, device.alertsCount)
	if device.mostFailingNeighbour == nil || device.alertsCount < neighbour.alertsCount {
		h.repository.updateMostFailingNeighbour(deviceName, neighbour)
	}
}

File: src/main/java/com/orange/craftsmanship/alarming/domain/AlertHandler.java
package com.orange.craftsmanship.alarming.domain;

public class AlertHandler implements NetworkSupervisionPort {

    private final DeviceRepositoryPort deviceRepositoryPort;

    public AlertHandler(DeviceRepositoryPort deviceRepositoryPort) {
        this.deviceRepositoryPort = deviceRepositoryPort;
    }

    @Override
    public void onNetworkInterfaceAlert(String sourceDeviceName, String targetDeviceName) {
        Device sourceDevice = deviceRepositoryPort.findDevice(sourceDeviceName).orElse(new Device(sourceDeviceName));
        sourceDevice.alertsCount += 1;
        Device targetDevice = deviceRepositoryPort.findDevice(targetDeviceName).orElse(new Device(targetDeviceName));
        targetDevice.alertsCount += 1;
        updateDevice(sourceDevice, targetDevice);
        updateDevice(targetDevice, sourceDevice);
    }

    private void updateDevice(Device device, Device otherDevice) {
        deviceRepositoryPort.updateAlertsCount(device.name(), device.alertsCount);
        if (!device.mostFailingNeighbour.isPresent() || device.mostFailingNeighbour.get().alertsCount < otherDevice.alertsCount) {
            deviceRepositoryPort.updateMostFailingNeighbour(device.name(), otherDevice);
        }
    }
}
